sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -0,0 +1,100 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "flash_ctrl_regs.h"  // Generated.
+
+static bool is_busy(const flash_ctrl_t *flash_ctrl) {
+  return mmio_region_get_bit32(flash_ctrl->base_addr,
+                               FLASH_CTRL_CTRL_REGWEN_REG_OFFSET,
+                               FLASH_CTRL_CTRL_REGWEN_EN_BIT) == 0;
+}
+
+void flash_ctrl_init(const flash_ctrl_t *flash_ctrl) {
+  // Initialize the flash controller.
+  mmio_region_write32(flash_ctrl->base_addr, FLASH_CTRL_INIT_REG_OFFSET,
+                      bitfield_bit32_write(0u, FLASH_CTRL_INIT_VAL_BIT, true));
+}
+
+flash_ctrl_status_t flash_ctrl_status_get(const flash_ctrl_t *flash_ctrl);
+
+rom_error_t flash_ctrl_read_start(flash_ctrl_t *flash_ctrl, uint32_t addr,
+                                  uint32_t word_count,
+                                  flash_ctrl_region_select_t region) {
+  if (is_busy(flash_ctrl)) {
+    return kErrorFlashCtrlBusy;
+  }
+
+  // Set the address.
+  mmio_region_write32(flash_ctrl->base_addr, FLASH_CTRL_ADDR_REG_OFFSET, addr);
+
+  // Set the operation of the transaction: read, program, or erase.
+  uint32_t control_reg_val = bitfield_field32_write(
+      0, FLASH_CTRL_CONTROL_OP_FIELD, FLASH_CTRL_CONTROL_OP_VALUE_READ);
+
+  // Set the partition to info here, and set it back bellow if we're accessing a
+  // data partition.
+  control_reg_val = bitfield_bit32_write(
+      control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 1u);
+
+  // Set the partition and info type.
+  switch (region) {
+    case kFlashCtrlRegionData:
+      // 0 indicates data partition.
+      control_reg_val = bitfield_bit32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 0u);
+      break;
+    case kFlashCtrlRegionInfo0:
+      // The partition is already set to info, so just set the info type.
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 0u);
+      break;
+    case kFlashCtrlRegionInfo1:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 1u);
+      break;
+    case kFlashCtrlRegionInfo2:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 2u);
```

Comment:
This is ok for now, but we should consider if there is a way to be clever with the values of flash_control_region_select_t so as to avoid having to have this switch/case translator.

IOW, could we write this bit of code as something like:
```
  control_reg_val |= (uint32_t)region
```

Reply:
@jon-flatley PTAL